src/main/java/com/jwplayer/southpaw/state/RocksDBState.java

```diff
@@ -243,46 +252,29 @@ public RocksDBState(Map<String, Object> config) {
     public void backup() {
         logger.info("Backing up RocksDB state");
         try {
-            switch(backupURI.getScheme().toLowerCase()) {
-                case FileHelper.SCHEME:
-                    backup(backupURI.getPath());
-                    break;
-                case S3Helper.SCHEME:
-                    String localBackupPath = getLocalBackupPath(uri);
-                    backup(localBackupPath);
-                    s3Helper.syncToS3(new URI(localBackupPath), backupURI);
-                    break;
-                default:
-                    throw new RuntimeException("Unsupported schema: " + backupURI.getScheme());
-            }
-        } catch(InterruptedException | ExecutionException | URISyntaxException | RocksDBException ex) {
+            openBackupEngine();
+            backupEngine.createNewBackup(rocksDB, true);
+            backupEngine.purgeOldBackups(backupsToKeep);
+        } catch(RocksDBException ex) {
             throw new RuntimeException(ex);
         }
-        logger.info("RocksDB state backup complete");
-    }
 
-    /**
-     * Backups the DB to a local path.
-     * @param backupPath - The local backup path
-     */
-    protected void backup(String backupPath) throws RocksDBException {
-        File file = new File(backupPath);
-        if(!file.exists()) file.mkdir();
-        logger.info("Opening RocksDB backup engine");
-        try (final BackupableDBOptions backupOptions = new BackupableDBOptions(backupPath)
-                .setShareTableFiles(true)
-                .setMaxBackgroundOperations(parallelism);
-            final BackupEngine backupEngine = BackupEngine.open(Env.getDefault(), backupOptions)) {
-            backupEngine.createNewBackup(rocksDB, true);
-            backupEngine.purgeOldBackups(backupsToKeep);
+        if(s3Helper != null) {
```

Comment:
Using `s3Helper` here and elsewhere as the signal that the backup location is ultimately in S3 feels weird to me. Partly because what this is really saying is "is the backup URI in s3?" but with a bit of obfuscation with whether `s3Helper` is set or not. And partly because S3 is singled out as special vs other storage locations. But S3 *is* special versus the local file system given how our version of RocksDB works and we're probably never going to support S3 like file storage of other cloud providers. Ideally supporting different file systems would be handled in RocksDB (I'm still curious how Flink does it) or with some level of abstraction here, but that seems overkill at this point.

Anywho, lots of words over something pretty trivial. :P

Reply:
Totally agree. Not even fully sure direct S3 uploads from within southpaw are something we really need/want long term anymore since for at least our use case we have persistent mounted volumes we save to now. I could see a world where we just handle persisting backups outside of the scope of southpaw. 

Can also see a world where we better define the BaseState interface as well as potentially a BaseStateBackup class which would solely handle backups. We could Then have something like a RocksDbStateBackup class that just backs up to a local filesystem and potentially extend that for S3 or other external persistence in the future.

Getting back to this PR change though, the change here isn't really changing how we determine if we're backing up to S3. It's still primarily being determined by the backup URI being `s3://` vs file://`. The main change is this is now determined in one place during `RocksDBState.configuration()` vs each time `RocksDBState.backup()` or `RocksDBState.restore()` is called. Is this something that we feel comfortable fixing in a future PR?